haystack/preview/components/reader.py

```diff
@@ -0,0 +1,239 @@
+        exp_logits /= exp_sum_per_query.unsqueeze(-1).unsqueeze(-1)
+        return exp_logits
+
+    def _postprocess(
+        self,
+        output,
+        sequence_ids: torch.Tensor,
+        attention_mask: torch.Tensor,
+        top_k: int,
+        encodings: List[Encoding],
+        query_ids: List[int],
+    ) -> Tuple[List[List[int]], List[List[int]], List[List[float]]]:
+        start = output.start_logits
+        end = output.end_logits
+        mask = sequence_ids == 1
+        mask[..., 0] = True  # type: ignore # mypy doesn't understand broadcasting
+        mask = torch.logical_and(mask, attention_mask == 1)  # type: ignore # see above
+        start = torch.where(mask, start, -torch.inf)  # type: ignore # see above
+        end = torch.where(mask, end, -torch.inf)  # type: ignore # see above
+        start = start.unsqueeze(-1)
+        end = end.unsqueeze(-2)
+
+        logits = start + end  # shape: (batch_size, seq_length (start), seq_length (end))
+        mask = torch.ones(logits.shape[-2:], dtype=bool)
+        mask = torch.triu(mask)  # End shouldn't be before start
+        mask[0, :1] = False
+        masked_logits = torch.where(mask, logits, -torch.inf)
+        probabilities = self._softmax_across_documents(masked_logits, query_ids)
+
+        flat_probabilities = probabilities.flatten(-2, -1)  # necessary for topk
+        candidates = torch.topk(flat_probabilities, top_k)
+        seq_length = logits.shape[-1]
+        start_candidates = candidates.indices // seq_length  # Recover indices from flattening
+        end_candidates = candidates.indices % seq_length
+        start_candidates = start_candidates.cpu()
+        end_candidates = end_candidates.cpu()
+
+        start_candidates = [
+            [encoding.token_to_chars(start)[0] if start != 0 else 0 for start in candidates]
+            for candidates, encoding in zip(start_candidates, encodings)
+        ]
+        end_candidates = [
+            [encoding.token_to_chars(end)[1] if end != 0 else 0 for end in candidates]
+            for candidates, encoding in zip(end_candidates, encodings)
+        ]
+        probabilities = candidates.values.cpu().tolist()
+
+        return start_candidates, end_candidates, probabilities
+
+    def _unflatten(
+        self,
+        start: List[List[int]],
+        end: List[List[int]],
+        probabilities: List[List[float]],
+        flattened_documents: List[str],
+        documents: List[List[Document]],
+        top_k: int,
+        query_ids: List[int],
+        document_ids: List[int],
+    ) -> List[List[Answer]]:
+        answers = [
+            (flattened_documents[document_id][start:end], probability)
+            for document_id, start_candidates_, end_candidates_, probabilities_ in zip(
+                document_ids, start, end, probabilities
+            )
+            for start, end, probability in zip(start_candidates_, end_candidates_, probabilities_)
+        ]
+
+        i = 0
```

Comment:
Nitpick: let's give it a better name. Like `answers_count_for_query` maybe? I think I understand what it does but you surely have a better name for it

Reply:
Ok, changed it